Strict redirect URL enforcement (#2917)
kanidm/kanidm: book/src/integrations/oauth2.md

members of `system_admins` or `idm_hp_oauth2_manage_priv` are able to create or manage OAuth2 client
integrations.

> You can create a new client by specifying it's client name, application display name and the landing

Reviewer: ```suggestion
You can create a new client by specifying its client name, application display name and the landing
```

> page (home page) of the client. The landing page is where users will be redirect to in order to

Reviewer 2: "redirected"

> page (home page) of the client. The landing page is where users will be redirect to in order to

Reviewer: ```suggestion
page (home page) of the client. The landing page is where users will be redirected to in order to
```

> begin an oauth2 flow.

Reviewer 2: Is that only used for the kanidm's dashboard then? Because generally I'd expect the client to initiate the flow. If it's the case, I suggest rephrasing this to point this exact usage.

Author: Clarified.

```bash
kanidm system oauth2 create <name> <displayname> <landing page url>
kanidm system oauth2 create nextcloud "Nextcloud Production" https://nextcloud.example.com
```

> You must now configure the redirect URL where the application expects oauth2 requests to be sent.

Reviewer: ```suggestion
You must now configure the redirect URL where the application expects OAuth2 requests to be sent.
```

kanidm system oauth2 enable-localhost-redirects mywebapp
```

## Alternate Redirect URLs

Some services may have a website URL as well as native applications. These native applications
require alternate redirection URLs to be configured so that after an OAuth2 exchange, the system can
redirect to the native application.

To support this Kanidm allows supplemental opaque origins to be configured on clients.

> {{#template ../templates/kani-warning.md imagepath=../images title=WARNING text=The ability to

Reviewer: Can't line wrap these 'cause it breaks the template, have to put the deno exception tags in ... found somewhere else

configure multiple redirect URLs is NOT intended to allow you to share a single Kanidm client
definition between multiple OAuth2 clients. This fundamentally breaks the OAuth2 security model and
is NOT SUPPORTED as a configuration. Multiple origins is only to allow supplemental redirects within
the _same_ client application. }}

```bash
kanidm system oauth2 add-redirect-url <name> <url>
kanidm system oauth2 remove-redirect-url <name> <url>

kanidm system oauth2 add-redirect-url nextcloud app://ios-nextcloud
```

Supplemental URLs are shown in the OAuth2 client configuration in the `oauth2_rs_origin` attribute.

### Strict Redirect URLs

Kanidm previously enforce that redirection targets only matched by _origin_ not the full URL. In

> 1.4.0 these URLs well expect a full URL match instead.

Reviewer: ```suggestion
1.4.0 these URLs follow the RFC and expect a full URL match instead.
```

kanidm system oauth2 enable-localhost-redirects mywebapp
```

## Alternate Redirect URLs

{{#template ../templates/kani-warning.md imagepath=../images title=WARNING text=SECURITY RISK }}

> > You **MUST NOT** shared a single OAuth2 client definition between multiple applications.

Reviewer 2: ```suggestion
> You **MUST NOT** share a single OAuth2 client definition between multiple applications.
```

>
> The ability to configure multiple redirect URLs is **NOT** to allow you to share a single Kanidm
> client definition between multiple OAuth2 clients.
>
> Sharing OAuth2 client configurations between applications **FUNDAMENTALLY BREAKS** the OAuth2
> security model and is **NOT SUPPORTED** as a configuration. The Kanidm Project **WILL NOT**
> support you if you attempt this.
>
> Multiple origins are **ONLY** to allow supplemental redirects within the _same_ client
> application.

Some services may have a website URL as well as native applications with opaque origins. These
native applications require alternate redirection URLs to be configured so that after an OAuth2
exchange, the system can redirect to the native application.

To support this Kanidm allows supplemental opaque origins to be configured on clients.

```bash
kanidm system oauth2 add-redirect-url <name> <url>
kanidm system oauth2 remove-redirect-url <name> <url>

kanidm system oauth2 add-redirect-url nextcloud app://ios-nextcloud
```

Supplemental URLs are shown in the OAuth2 client configuration in the `oauth2_rs_origin` attribute.

### Strict Redirect URLs

> Kanidm previously enforce that redirection targets only matched by _origin_ not the full URL. In

Reviewer 2: ```suggestion
Kanidm previously enforce that redirection targets only matched by _origin_, not the full URL. In
```
